lib/semantics/canonicalize-do.cc

```diff
@@ -17,103 +17,72 @@
-  bool Pre(Block &block) { return VisitBlock(block); }
-  template<typename T> bool Pre(Statement<T> &statement) {
-    if (!labelInfos_.empty() && statement.label.has_value() &&
-        labelInfos_.back().label == *statement.label) {
-      auto currentLabel{labelInfos_.back().label};
-      if constexpr (std::is_same_v<T, common::Indirection<EndDoStmt>>) {
-        std::get<ExecutableConstruct>(currentPosition_.iter->u).u =
-            Statement<ActionStmt>{
-                std::optional<Label>{currentLabel}, ContinueStmt{}};
+  void Post(Block &block) {
+    std::vector<LabelInfo> stack;
+    for (auto i{block.begin()}, end{block.end()}; i != end; ++i) {
+      if (auto *executableConstruct{std::get_if<ExecutableConstruct>(&i->u)}) {
+        std::visit(
+            common::visitors{[](auto &) {},
+                [&](Statement<common::Indirection<LabelDoStmt>> &labelDoStmt) {
+                  auto &label{std::get<Label>(labelDoStmt.statement->t)};
+                  stack.push_back(LabelInfo{i, label});
+                },
+                [&](Statement<common::Indirection<EndDoStmt>> &endDoStmt) {
+                  CanonicalizeIfMatch(block, stack, i, endDoStmt);
+                },
+                [&](Statement<ActionStmt> &actionStmt) {
+                  CanonicalizeIfMatch(block, stack, i, actionStmt);
+                }},
+            executableConstruct->u);
       }
-      do {
-        currentPosition_.iter = MakeCanonicalForm(*currentPosition_.block,
-            labelInfos_.back().iter, currentPosition_.iter);
-        labelInfos_.pop_back();
-      } while (
-          !labelInfos_.empty() && labelInfos_.back().label == currentLabel);
     }
-    return false;
   }
 
 private:
-  bool VisitBlock(Block &block) {
-    CanonicalizationOfDoLoops canonicalizationOfDoLoops{labelInfos_};
-    canonicalizationOfDoLoops.TraverseBlock(block);
-    return false;
-  }
-  void TraverseBlock(Block &block) {
-    const auto &endIter{block.end()};
-    currentPosition_.block = &block;
-    for (auto iter{block.begin()}; iter != endIter; ++iter) {
-      ExecutionPartConstruct &executionPartConstruct{*iter};
-      currentPosition_.iter = iter;
-      if (auto *executableConstruct{
-              std::get_if<ExecutableConstruct>(&executionPartConstruct.u)}) {
-        if (auto *labelDoLoop{
-                std::get_if<Statement<common::Indirection<LabelDoStmt>>>(
-                    &executableConstruct->u)}) {
-          labelInfos_.push_back(
-              LabelInfo{iter, std::get<Label>(labelDoLoop->statement->t)});
-        }
+  template<typename T>
+  void CanonicalizeIfMatch(Block &originalBlock, std::vector<LabelInfo> &stack,
+      Block::iterator &i, Statement<T> &statement) {
+    if (!stack.empty() && statement.label.has_value() &&
+        stack.back().label == *statement.label) {
+      auto currentLabel{stack.back().label};
+      if constexpr (std::is_same_v<T, common::Indirection<EndDoStmt>>) {
```

Comment:
An END DO statement should never be encountered in a Block unless it bears a label used by an earlier label DO in the same Block.  That should be CHECKed.